### PR TITLE
#92 C4 m0hamed s11 search by category

### PR DESCRIPTION
**Pull request by m0hamed**
@NourhanMohamed @mostafahassaan #92

### CODE REVIEW

File: arability/app/assets/javascripts/bootstrap-tagmanager.js [hunk 1]
```diff
@@ -0,0 +1,602 @@
+/* ===================================================
```
Comment (NourhanMohamed) [inline arability/app/assets/javascripts/bootstrap-tagmanager.js:1]: What have you changed in this file? I'm not reviewing it all :D

File: arability/app/controllers/application_controller.rb [hunk 2]
```diff
@@ -1,6 +1,5 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery
-  rescue_from Exception, :with => :error_render_method
```
Comment (NourhanMohamed) [inline arability/app/controllers/application_controller.rb:3]: r u sure u deleted this line?

File: arability/app/models/category.rb [hunk 3]
```diff
@@ -1,12 +1,16 @@
...
+  attr_accessible :english_name, :arabic_name
   has_and_belongs_to_many :keywords
   has_and_belongs_to_many :projects
 
-  validates_uniqueness_of :name
-  validates_presence_of :name
-  validates_format_of :name, :with => /^([\u0621-\u0652 ]+|[a-zA-z ]+)$/
+  validates_uniqueness_of :english_name
+  validates_presence_of :english_name
+  validates_format_of :english_name, :with => /^([a-zA-Z ]+)$/
```
Comment (NourhanMohamed) [inline arability/app/models/category.rb]: can u use with: /^([a-zA-Z ]+)$/
or will this cause problems?

File: arability/app/models/category.rb [hunk 4]
```diff
@@ -1,12 +1,16 @@
...
+
+  validates_uniqueness_of :arabic_name
+  validates_presence_of :arabic_name
+  validates_format_of :arabic_name, :with => /^([\u0621-\u0652 ]+)$/
```
Comment (NourhanMohamed) [inline arability/app/models/category.rb]: same as the previous comment

File: arability/app/views/search/search_keywords.html.erb [hunk 5]
```diff
@@ -35,3 +36,4 @@
 		<% end %>
 	</div>
 </div>
+<%= render 'categories-autocomplete' %>
```
Comment (NourhanMohamed) [inline arability/app/views/search/search_keywords.html.erb]: use double quotes rather than single quotes

File: arability/lib/string_helper.rb [hunk 6]
```diff
@@ -8,7 +8,7 @@ module StringHelper
   def is_english_string(text)
     text = text.strip
     text = text.split(" ").join("")
-    if text.match /^[a-zA-Z]+$/
+    if text.match /^[a-zA-Z ]+$/
```
Comment (NourhanMohamed) [inline arability/lib/string_helper.rb]: use conditional

File: arability/lib/tasks/populate.rake [hunk 7]
```diff
@@ -27,6 +27,12 @@ namespace :db do
     Keyword.create(:name => "إعدادات", :approved => true, :is_english => false)
     Keyword.create(:name => "انجازات", :approved => true, :is_english => false)
 
+    Category.delete_all
```
Comment (NourhanMohamed) [inline arability/lib/tasks/populate.rake:30]: remove => from subsequent lines

Comment (NourhanMohamed) [inline arability/lib/tasks/populate.rake:30]: and from previous ones :D

File: arability/app/assets/javascripts/bootstrap-tagmanager.js (hunk 1 above)
Comment (m0hamed) [inline arability/app/assets/javascripts/bootstrap-tagmanager.js:1]: As you can see its a plugin for bootstrap and a javascript file. Both of which imply that you shouldnt review it anyway.

File: arability/app/controllers/application_controller.rb (hunk 2 above)
Comment (m0hamed) [inline arability/app/controllers/application_controller.rb:3]: Yes we dont want it till the end just before the demo. We need to see the exceptions

File: arability/app/models/category.rb (hunk 3 above)
Comment (m0hamed) [inline arability/app/models/category.rb]: I think it would work, Let me try it

File: arability/lib/tasks/populate.rake (hunk 7 above)
Comment (m0hamed) [inline arability/lib/tasks/populate.rake:30]: this is the database seed. Why are you reviewing this

**Comment by m0hamed** [comment]
please review again

File: arability/app/models/category.rb [hunk 8]
```diff
@@ -17,15 +21,38 @@ class << self
...
+    arabic_name.strip!
+
+    english_name.downcase!
+
+    english_name = english_name.split(" ").join(" ")
+    arabic_name = arabic_name.split(" ").join(" ")
+
+    category = Category.where(english_name: english_name, arabic_name: arabic_name).first_or_create
+
     return category.save ? [true, category] : [false, category]
```
Comment (NourhanMohamed) [inline arability/app/models/category.rb]: remove redundant "return"

File: arability/app/views/search/_categories-autocomplete.html.erb [hunk 9]
```diff
@@ -0,0 +1,24 @@
+<script type="text/javascript">
+  <% categories = Category.all.map {|c| c.get_name_by_locale} %>
```
Comment (NourhanMohamed) [inline arability/app/views/search/_categories-autocomplete.html.erb]: leave a space before the } and after the {

File: arability/lib/tasks/populate.rake [hunk 10]
```diff
@@ -1,7 +1,8 @@
 #encoding:utf-8
+#%s/:\([a-z_]\+\) =>/\1:/g
```
Comment (NourhanMohamed) [inline arability/lib/tasks/populate.rake]: remove this comment

File: arability/spec/helpers/string_helper_spec.rb [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  it "Should not recognise arabic strings with spaces" do
+    is_english_string("سبت").should eq(false)
+  end
+
+  it "Should not recognise mixed strings with spaces" do
+    is_english_string("سبت akdjf").should eq(false)
+  end
+
```
Comment (NourhanMohamed) [inline arability/spec/helpers/string_helper_spec.rb:19]: remove extra line

Comment (NourhanMohamed) [inline arability/spec/helpers/string_helper_spec.rb:19]: unless u're intending this for git merge

File: arability/spec/models/keyword_spec.rb [hunk 12]
```diff
@@ -48,11 +48,11 @@
   end
 
   it "should add keywords with categories" do
-    success, category = Category.add_category_to_database_if_not_exists("test")
-    success, keyword = Keyword.add_keyword_to_database("test",true,true, ["test"])
+    success, category = Category.add_category_to_database_if_not_exists("test", "تشتبت")
+    success, keyword = Keyword.add_keyword_to_database("test",true,true, ["تشتبت"])
```
Comment (NourhanMohamed) [inline arability/spec/models/keyword_spec.rb]: spaces after the commas

File: arability/app/models/category.rb [hunk 13]
```diff
@@ -17,15 +21,38 @@ class << self
...
+  def self.add_category_to_database_if_not_exists(english_name, arabic_name)
+    english_name.strip!
+    arabic_name.strip!
+
+    english_name.downcase!
+
+    english_name = english_name.split(" ").join(" ")
+    arabic_name = arabic_name.split(" ").join(" ")
+
+    category = Category.where(english_name: english_name, arabic_name: arabic_name).first_or_create
```
Comment (NourhanMohamed) [inline arability/app/models/category.rb]: line exceeded 80 characters

File: arability/app/views/search/_categories-autocomplete.html.erb [hunk 14]
```diff
@@ -0,0 +1,24 @@
+<script type="text/javascript">
+  <% categories = Category.all.map { |c| c.get_name_by_locale} %>
```
Comment (NourhanMohamed) [inline arability/app/views/search/_categories-autocomplete.html.erb]: missing space before }

File: arability/spec/controllers/search_controller_spec.rb [hunk 15]
```diff
@@ -31,20 +31,18 @@
...
       gamer.date_of_birth = "1993-03-23"
       gamer.email = "nour@gmail.com"
       gamer.password = "1234567"
-      gamer.saveg = Gamer.new
+      gamer.save validate: false
     end
 
     it "should get only keywords in category" do
       c.keywords << k
       a = create_logged_in_developer()
```
Comment (NourhanMohamed) [inline arability/spec/controllers/search_controller_spec.rb]: remove unnecessary brackets in test cases

**Pull request merged**